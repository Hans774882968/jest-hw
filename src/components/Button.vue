<template>
  <button
    class="my-button"
    :class="computedClass"
    :type="nativeType"
    :disabled="disabled || loading"
    @click="handleClick"
  >
    <slot />
  </button>
</template>

<script lang="ts">
import { Vue, Prop, Component } from 'vue-property-decorator';

@Component
export default class Button extends Vue {
  @Prop({
    default: 'button',
  }) nativeType!: 'button' | 'submit' | 'reset' | undefined;

  @Prop({
    type: String,
    default: 'primary',
  }) type!: string;

  @Prop({
    type: String,
    default: 'medium',
  }) size!: string;

  @Prop({
    type: String,
    default: 'round',
  }) shape!: string;

  @Prop({
    type: Boolean,
    default: false,
  }) disabled!: boolean;

  @Prop({
    type: Boolean,
    default: false,
  }) loading!: boolean;

  get computedClass() {
    return {
      [`btn-type-${this.type}`]: !!this.type,
      [`btn-size-${this.size}`]: !!this.size,
      [`btn-shape-${this.shape}`]: !!this.shape,
      'btn-disabled': !!this.disabled || !!this.loading,
      'btn-loading': !!this.loading,
    };
  }

  handleClick(evt: MouseEvent) {
    this.$emit('click', evt);
  }
}
</script>
