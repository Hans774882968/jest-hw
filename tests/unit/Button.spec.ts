import { shallowMount, Wrapper } from '@vue/test-utils';
import Button from '@/components/Button.vue';

describe('Button.vue', () => {
  let wrapper: Wrapper<any, Element>;
  afterEach(() => {
    wrapper.destroy();
  });

  it('renders basic HTML structure properly', () => {
    wrapper = shallowMount(Button);
    expect(wrapper.element.nodeName).toBe('BUTTON');
    expect(wrapper.classes()).toContain('my-button');
  });

  // Or you can use snapshot to save many expects
  it('render element properly (judge by snapshot)', () => {
    wrapper = shallowMount(Button);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('slot', () => {
    wrapper = shallowMount(Button, {
      slots: {
        default: '<span class="acmer">hans</span>',
      },
    });
    const innerEl = wrapper.find('.acmer');
    expect(innerEl.element.nodeName).toBe('SPAN');
    expect(wrapper.text()).toMatch('hans');
  });

  it('nativeType', () => {
    wrapper = shallowMount(Button, {
      propsData: { nativeType: 'submit' },
    });
    expect(wrapper.attributes('type')).toBe('submit');
  });

  it('type', () => {
    wrapper = shallowMount(Button, {
      propsData: { type: 'danger' },
    });
    expect(wrapper.classes()).toContain('btn-type-danger');
  });

  it('size', () => {
    wrapper = shallowMount(Button, {
      propsData: { size: 'small' },
    });
    expect(wrapper.classes()).toContain('btn-size-small');
  });

  it('shape', () => {
    wrapper = shallowMount(Button, {
      propsData: { shape: 'round' },
    });
    expect(wrapper.classes()).toContain('btn-shape-round');
  });

  it('loading', () => {
    wrapper = shallowMount(Button, {
      propsData: { loading: true },
    });
    expect(wrapper.classes()).toContain('btn-loading');
  });

  // click tests
  it('click', () => {
    const clickHandler = jest.fn();
    wrapper = shallowMount(Button, {
      listeners: {
        click: clickHandler,
      },
    });
    for (let i = 0; i < 5; i += 1) wrapper.trigger('click');

    expect(clickHandler).toHaveBeenCalledTimes(5);
  });

  it('click inside slot', () => {
    const clickHandler = jest.fn();
    wrapper = shallowMount(Button, {
      slots: {
        default: '<span class="inner" />',
      },
      listeners: {
        click: clickHandler,
      },
    });

    const innerEl = wrapper.find('.inner');
    for (let i = 0; i < 5; i += 1) innerEl.trigger('click');
    expect(clickHandler).toHaveBeenCalledTimes(5);
  });

  it('disabled', () => {
    const clickHandler = jest.fn();
    wrapper = shallowMount(Button, {
      propsData: { disabled: true },
      listeners: {
        click: clickHandler,
      },
    });
    expect(wrapper.classes()).toContain('btn-disabled');
    for (let i = 0; i < 5; i += 1) wrapper.trigger('click');
    expect(clickHandler).not.toHaveBeenCalled();
  });

  it('disabled when loading', () => {
    const clickHandler = jest.fn();
    wrapper = shallowMount(Button, {
      propsData: { loading: true },
      listeners: {
        click: clickHandler,
      },
    });
    for (let i = 0; i < 5; i += 1) wrapper.trigger('click');
    expect(clickHandler).not.toHaveBeenCalled();
  });
});
