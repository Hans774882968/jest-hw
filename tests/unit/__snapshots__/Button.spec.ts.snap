// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue render element properly (judge by snapshot) 1`] = `
<button
  class="my-button btn-type-primary btn-size-medium btn-shape-round"
  type="button"
/>
`;
